source/val/validate_instruction.cpp

```diff
@@ -506,6 +506,24 @@ spv_result_t InstructionPass(ValidationState_t& _, const Instruction* inst) {
                   "outside of a function";
       }
     }
+    if (_.IncludesIntOrFloatOfWidth(inst->type_id(), SpvOpTypeInt, 8) &&
```

Comment:
These checks belong in ValidateVariable in validate_memory.cpp; however, they seem incorrect. `OpCapability Int8` should allow full use of 8-bit integers. The checks seem too strict. 

Code exists in the validation state to control whether int8 or float16 can be declared and/or used. You should probably expand on that.

Reply:
Hi, thank you very much for looking into this. I am looking for a way of validating the following rule in the spec:

>`ifdef::VK_KHR_shader_float16_int8+!VK_KHR_8bit_storage[]`
Any OpVariable with the result type pointing to an 8-bit integer object or an object containing an 8-bit integer element must: not have one of the following as its Storage Class operand:
Uniform
PushConstant
`ifdef::VK_VERSION_1_1,VK_KHR_storage_buffer_storage_class[]`
StorageBuffer
`endif::VK_VERSION_1_1,VK_KHR_storage_buffer_storage_class[]
endif::VK_KHR_shader_float16_int8+!VK_KHR_8bit_storage[]`

Therefore my first guess was that my check would have to look something like:

```
if (_.IncludesIntOrFloatOfWidth(inst->type_id(), SpvOpTypeInt, 8) &&
*** something *** &&
(storage_class == SpvStorageClassUniform ||
storage_class == SpvStorageClassPushConstant ||
(_.HasCapability(SpvCapabilityStorageBuffer8BitAccess) &&
storage_class == SpvStorageClassStorageBuffer))) {
return _.diag(SPV_ERROR_INVALID_LAYOUT, inst)
<< "Variables can not point to an 8-bit integer "
"without 8-bit storage capabilities";
}
```

However I am not sure what this `something` should be. Originally I thought I had to check for what Vulkan extensions are enabled, but this is not possible from SPIRV-Tools?

Reply:
No, the validator does not know what Vulkan extensions are enabled in general. There is enough information in SPIR-V to determine what is allowable in most cases. The `something` in your first version was too strict because the `Int8` capability allows full use of 8-bit integers.

Reply:
> Code exists in the validation state to control whether int8 or float16 can be declared and/or used. You should probably expand on that.

I found
```
  switch (cap) {
...
    case SpvCapabilityInt8:
      features_.use_int8_type = true;
      features_.declare_int8_type = true;
      break;
```
But not sure how to implement this validation here, since we need to validate the StorageClass of each OpVariable.

>These checks belong in ValidateVariable in validate_memory.cpp; however, they seem incorrect. OpCapability Int8 should allow full use of 8-bit integers. The checks seem too strict.

I moved the checks to validate_memory.cpp removing the check for Int8 capability, but still check that 8bit_storate is _not_ enabled.